packages/teams-js/src/private/externalAppCommands.ts

```diff
@@ -0,0 +1,192 @@
+import { errorNotSupportedOnPlatform } from '../public/constants';
+import { runtime } from '../public/runtime';
+
+/**
+ * v2 APIs telemetry file: All of APIs in this capability file should send out API version v2 ONLY
+ */
+const externalAppCommandsTelemetryVersionNumber: ApiVersionNumber = ApiVersionNumber.V_2;
+
+/**
+ * @hidden
+ * Namespace to delegate the ActionCommand to the host
+ * @internal
+ * Limited to Microsoft-internal use
+ *
+ * @beta
+ */
+export namespace externalAppCommands {
+  /**
+   * @hidden
+   * The payload of IActionCommandResponse
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface IActionCommandResponse {
+    /**
+     * type will be result or message
+     */
+    type?: ResultTypes | null;
+    /**
+     * attachment items will exist only when type is result
+     */
+    items?: MessageExtensionItem[] | null;
+    /**
+     * text will exist only when type is message
+     */
+    text?: string | null;
+  }
+
+  /**
+   * @hidden
+   * The result type for the ActionCommandResponse
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type ResultTypes = 'result' | 'message';
+
+  /**
+   * @hidden
+   * The payload of attachment item
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface MessageExtensionItem {
+    card: Card;
+    previewCard?: Card | null;
+  }
+
+  /**
+   * @hidden
+   * The payload of Card
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface Card {
+    appIcon?: string | null;
+    appId?: string | null;
+    appName?: string | null;
+    cardClientId?: string | null;
+    content: string;
+    contentType: string;
+    locale?: string | null;
+    name?: string | null;
+    userSpecificContent?: UserSpecificCardContent | null;
+    sourceType?: CardSourceType | null;
+  }
+
+  /**
+   * @hidden
+   * The payload that is used when executing an Adaptive Card Action.Submit
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface UserSpecificCardContent {
+    content?: string | null;
+    timeStamp: number;
+    version?: string | null;
+    authBlockLastUsedOn?: number | null;
+  }
+
+  /**
+   * @hidden
+   * The payload that is used for CardSourceType
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export type CardSourceType =
+    | 'ZeroInstallAppInstalledAppLinkUnfurl'
+    | 'NonZeroInstallAppInstalledAppLinkUnfurl'
+    | 'ZeroInstallAppNotInstalledAppLinkUnfurl'
+    | 'URLP'
+    | 'URLPRichInfo'
+    | 'NonLinkUnfurling';
+
+  /**
+   *
+   * @hidden
+   * Error that can be thrown from IExternalAppCommandsService.processActionCommand
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   *
+   * @beta
+   */
+  export interface ActionCommandError {
+    errorCode: string;
```

Comment:
>string [](http://example.com/codeflow?start=15&length=6)

An error "code" sounds like a number. Is it really a string?

Reply:
That's a great question, and the errorCode is string in the design docs: https://microsoft-my.sharepoint-df.com/:w:/r/personal/erinha_microsoft_com/_layouts/15/Doc.aspx?sourcedoc=%7Be3ed58ec-b5d5-4c11-8942-182bba9a0c7e%7D&action=edit&wdPid=17fd8a17&share=IQHsWO3j1bURTIlCGCu6mgx-AWrjzLcPCv_GPhxBpYPA1Iw @lakhveerk could you help confirming the type of the errorCode for ActionCommandError? Thanks a lot!

Reply:
Talked with @lakhveerk offline, `Error code is INTERNAL_ERROR for everything and yes, it is string type. It just followed a pattern that we had for Cards.` 
And I updated with a generic errorCode type named `ExternalAppErrorCode` in constants file for externalAppCommands api and external-app-card for Action Submit specifically.